sim/firesim-lib/src/main/scala/bridges/TerminationBridge.scala

```diff
@@ -0,0 +1,112 @@
+  * @param isError When true, instructs the driver to return a non-zero value when exiting under this condition.
+**/
+case class TerminationCondition(isErr: Boolean, message : String)
+
+//A seq of termination conditions one for each concurrently running program
+case class TerminationBridgeParams(
+  codeMessage: Seq[TerminationCondition]
+)                
+
+class TerminationBridgeTargetIO(params: TerminationBridgeParams) extends Bundle {
+  val clock = Input(Clock())
+  val errCode = Input(UInt((log2Ceil((params.codeMessage).size)).W))
+  val valid = Input(Bool()) 
+}
+
+class TerminationBridgeHostIO(params: TerminationBridgeParams)
+                              (private val targetIO : TerminationBridgeTargetIO = new TerminationBridgeTargetIO(params)) 
+                               extends Bundle with ChannelizedHostPortIO{
+  def targetClockRef = targetIO.clock 
+  //It is just inputs from the target to indicate test completion
+  //The HostPort annotation would not work with aggregates so creating individual channels.
+  val errCodeH = InputChannel(targetIO.errCode)
+  val validH = InputChannel(targetIO.valid)
+}
+
+class TerminationBridge(params: TerminationBridgeParams) extends BlackBox
+    with Bridge[TerminationBridgeHostIO, TerminationBridgeModule] {
+  val io = IO(new TerminationBridgeTargetIO(params))
+  val bridgeIO = new TerminationBridgeHostIO(params)(io)
+
+  val constructorArg = Some(params)
+
+  generateAnnotations()
+}
+
+object TerminationBridgeWithCode {
+  def apply(clock: Clock, finish: Bool, errCode: UInt, params: TerminationBridgeParams): TerminationBridge = {
+    val tbm = Module(new TerminationBridge(params))
+    tbm.io.errCode := errCode
+    tbm.io.valid := finish
+    tbm.io.clock := clock 
+    tbm  
+  }
+}
+
+object TerminationBridge {
+  def apply(clock: Clock, finish: Seq[Bool], errCode: Seq[TerminationCondition], params: TerminationBridgeParams): TerminationBridge = {
+    val finishAll = finish.reduce(_||_)
+    val errMessageID = PriorityEncoder((errCode.map(x => (x.isErr).B)))
+    val tfm = TerminationBridgeWithCode(clock, finishAll, errMessageID, params)
+    tfm  
+  }
+}
+
+
+class TerminationBridgeModule(params: TerminationBridgeParams)(implicit p: Parameters) extends BridgeModule[TerminationBridgeHostIO]()(p) {
+  lazy val module = new BridgeModuleImp(this) {
+    // This creates the interfaces for all of the host-side transport
+    // AXI4-lite for the simulation control bus, =
+    // AXI4 for DMA
+    val io = IO(new WidgetIO())
+
+
+
+    // This creates the host-side interface of your TargetIO
+    val hPort = IO(new TerminationBridgeHostIO(params)())
+    
+    //Resume the simulation after the target called for termination
+    val resume = WireInit(false.B)
+
+    val statusDone = Module(new Queue(Bool(), 1))
+    val errCode = Module(new Queue(UInt(log2Ceil((params.codeMessage).size).W), 1))
```

Comment:
A DRY-er way to do this: 
```suggestion
    val errCode = Module(new Queue(hPort.errCodeH.bits.cloneType), 1))
```

Also there is a "Queue" companion object that more simply does this that you should have a look at: 
https://www.chisel-lang.org/api/latest/chisel3/util/Queue$.html